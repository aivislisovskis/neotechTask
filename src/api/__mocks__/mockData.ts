import { ApiDataRow } from '../../types';

export const listItems: ApiDataRow[] = [
  {
    _id: 'idofentry1',
    name: 'Andris',
    surname: 'Berzins',
    age: 20,
    email: 'some@email.com',
    company: 'Name Of Company',
  },
  {
    _id: 'idofentry2',
    name: 'Janis',
    surname: 'Berzins',
    age: 20,
    email: 'some@email.com',
    company: 'Name Of Company',
  },
  {
    _id: 'idofentry3',
    name: 'Valis',
    surname: 'Berzins',
    age: 20,
    email: 'some@email.com',
    company: 'Name Of Company',
  },
  {
    _id: 'idofentry4',
    name: 'Zanis',
    surname: 'Berzins',
    age: 20,
    email: 'some@email.com',
    company: 'Name Of Company',
  }
];

export const singleNewItem: ApiDataRow = {
  name: 'Janis',
  surname: 'Berzins',
  age: 44,
  email: 'some@email.com',
  company: 'Name Of Company',
};