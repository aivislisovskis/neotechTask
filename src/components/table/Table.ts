import { create } from '../../helpers/create';
import { Elements } from '../../helpers/create.types';
import { RowData, TableData, TableOptions } from './table.types';
import styles from './table.css';
import { Row } from './Row';

export class Table {
  body: HTMLElement | null = null;

  constructor(public options: TableOptions) {
    this.body = create(Elements.div, {className: styles.body, content: [this.createHeader({ data: options.headers, id: null }), ...this.prepareInitData()]});
  }

  private createHeader(rowData: RowData): HTMLElement | null {
    if (this.options) {
      const row = new Row({ data: rowData, header: true});
      return row.body;
    }

    return null;
  }

  private prepareInitData(): (HTMLElement | null)[] {
    if (this.options?.data) {
      return this.options.data.map((row: RowData): HTMLElement | null => {
        return this.createRow(row);
      });
    }

    return [];
  }

  private createRow(row: RowData) {
    const rowElement = new Row({ data: row });
    return rowElement.body || null;
  }

  public add(data: TableData) {
      data.forEach((row: RowData) => {
        const rowElement = this.createRow(row);
        rowElement && this.body && this.body.appendChild(rowElement);
      })
  }
}
