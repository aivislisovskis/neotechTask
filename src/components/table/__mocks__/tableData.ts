import { ColumnDefinition, TableData } from '../table.types';
import { ColumnType } from '../../../elements/input/input.types';

export const tableData: TableData = [
  {
    data: ['Janis', 'Berzins', '10', 'mail@mail.ll', 'Some company'],
    id: 'someId',
  },
  {
    data: ['Peteris', 'Klavins', '20', 'rail@mail.ll', 'One company'],
    id: 'someId2',
  },
  {
    data: ['Zane', 'Kocina', '25', 'nail@mail.ll', 'Second company'],
    id: 'someId3',
  },
];

export const tableColumns: ColumnDefinition[] = [
  {
    header: 'name',
    type: ColumnType.string,
  },
  {
    header: 'surname',
    type: ColumnType.string,
  },
  {
    header: 'age',
    type: ColumnType.number,
  },
  {
    header: 'email',
    type: ColumnType.string,
  },
  {
    header: 'company',
    type: ColumnType.string,
  },
  {
    header: 'actions',
    type: ColumnType.button,
  }
];