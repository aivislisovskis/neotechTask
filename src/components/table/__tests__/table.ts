import { Table } from '../Table';
import { tableColumns, tableData } from '../__mocks__/tableData';

describe('Table > Table', () => {
  const onNewHandler = jest.fn();

  const tableOptions = {
    data: tableData,
    headers: tableColumns,
    onNew: onNewHandler,
  };

  const mouseEvent = new MouseEvent('click', {
    button: 0,
    buttons: 1,
  });

  it('*Snapshot, simple render', () => {
    const table = new Table(tableOptions);

    expect(table.body?.childNodes.length).toBe(5);
    expect(table.body).toMatchSnapshot();
  });

  it('*Snapshot, simple render, no initial data', () => {
    const tableOptionsBlank = { ...tableOptions };
    delete(tableOptionsBlank.data);

    const table = new Table(tableOptionsBlank);

    expect(table.body?.childNodes.length).toBe(2);
    expect(table.body).toMatchSnapshot();
  });

  it('Trigger new row action' , () => {
    const table = new Table(tableOptions);

    table.body?.firstChild?.firstChild?.dispatchEvent(mouseEvent);
    expect(onNewHandler).toBeCalled();
  });

  it('*Check callback for delete row', () => {
    const table = new Table(tableOptions);

    tableOptions.data[0].id && table.deleteCallback(tableOptions.data[0].id);

    expect(table.body?.childNodes.length).toBe(4);
    expect(table.body).toMatchSnapshot();
  });

  it('*Check callback for delete row, incorrect id', () => {
    const table = new Table(tableOptions);

    tableOptions.data[0].id && table.deleteCallback('incorrectId');

    expect(table.body?.childNodes.length).toBe(5);
    expect(table.body).toMatchSnapshot();
  });

  it('*Check callback for add new', () => {
    const table = new Table(tableOptions);

    table.newCallback(tableOptions.data[0]);

    expect(table.body?.childNodes.length).toBe(6);
    expect(table.body).toMatchSnapshot();
  });

  it('*Check callback for update data', () => {
    const table = new Table(tableOptions);

    const updatedData = {
      data: ['New name', 'New surname', '33', 'new@mail.co', 'New company'],
      id: 'someId'
    };

    tableOptions.data[0].id && table.updateItem(updatedData);

    expect(table.body).toMatchSnapshot();
  });

  it('*Check callback for update data, incorrect id', () => {
    const table = new Table(tableOptions);

    const updatedData = {
      data: ['New name', 'New surname', '33', 'new@mail.co', 'New company'],
      id: 'incorrectId'
    };

    tableOptions.data[0].id && table.updateItem(updatedData);

    expect(table.body).toMatchSnapshot();
  });

  it('*Check callback for add list of new data', () => {
    const table = new Table(tableOptions);

    tableOptions.data[0].id && table.add(tableOptions.data);

    expect(table.body?.childNodes.length).toBe(8);
    expect(table.body).toMatchSnapshot();
  });
});