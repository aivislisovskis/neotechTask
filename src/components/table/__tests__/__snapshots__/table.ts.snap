// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table > Table *Check callback for add list of new data 1`] = `
<div>
  <div>
    <button>
      Add New Row
    </button>
  </div>
  <div>
    <div>
      name
    </div>
    <div>
      surname
    </div>
    <div>
      age
    </div>
    <div>
      email
    </div>
    <div>
      company
    </div>
    <div>
      actions
    </div>
  </div>
  <div>
    <div>
      Janis
    </div>
    <div>
      Berzins
    </div>
    <div>
      10
    </div>
    <div>
      mail@mail.ll
    </div>
    <div>
      Some company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Peteris
    </div>
    <div>
      Klavins
    </div>
    <div>
      20
    </div>
    <div>
      rail@mail.ll
    </div>
    <div>
      One company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Zane
    </div>
    <div>
      Kocina
    </div>
    <div>
      25
    </div>
    <div>
      nail@mail.ll
    </div>
    <div>
      Second company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Janis
    </div>
    <div>
      Berzins
    </div>
    <div>
      10
    </div>
    <div>
      mail@mail.ll
    </div>
    <div>
      Some company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Peteris
    </div>
    <div>
      Klavins
    </div>
    <div>
      20
    </div>
    <div>
      rail@mail.ll
    </div>
    <div>
      One company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Zane
    </div>
    <div>
      Kocina
    </div>
    <div>
      25
    </div>
    <div>
      nail@mail.ll
    </div>
    <div>
      Second company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`Table > Table *Check callback for add new 1`] = `
<div>
  <div>
    <button>
      Add New Row
    </button>
  </div>
  <div>
    <div>
      name
    </div>
    <div>
      surname
    </div>
    <div>
      age
    </div>
    <div>
      email
    </div>
    <div>
      company
    </div>
    <div>
      actions
    </div>
  </div>
  <div>
    <div>
      Janis
    </div>
    <div>
      Berzins
    </div>
    <div>
      10
    </div>
    <div>
      mail@mail.ll
    </div>
    <div>
      Some company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Peteris
    </div>
    <div>
      Klavins
    </div>
    <div>
      20
    </div>
    <div>
      rail@mail.ll
    </div>
    <div>
      One company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Zane
    </div>
    <div>
      Kocina
    </div>
    <div>
      25
    </div>
    <div>
      nail@mail.ll
    </div>
    <div>
      Second company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Janis
    </div>
    <div>
      Berzins
    </div>
    <div>
      10
    </div>
    <div>
      mail@mail.ll
    </div>
    <div>
      Some company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`Table > Table *Check callback for delete row 1`] = `
<div>
  <div>
    <button>
      Add New Row
    </button>
  </div>
  <div>
    <div>
      name
    </div>
    <div>
      surname
    </div>
    <div>
      age
    </div>
    <div>
      email
    </div>
    <div>
      company
    </div>
    <div>
      actions
    </div>
  </div>
  <div>
    <div>
      Peteris
    </div>
    <div>
      Klavins
    </div>
    <div>
      20
    </div>
    <div>
      rail@mail.ll
    </div>
    <div>
      One company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Zane
    </div>
    <div>
      Kocina
    </div>
    <div>
      25
    </div>
    <div>
      nail@mail.ll
    </div>
    <div>
      Second company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`Table > Table *Check callback for delete row, incorrect id 1`] = `
<div>
  <div>
    <button>
      Add New Row
    </button>
  </div>
  <div>
    <div>
      name
    </div>
    <div>
      surname
    </div>
    <div>
      age
    </div>
    <div>
      email
    </div>
    <div>
      company
    </div>
    <div>
      actions
    </div>
  </div>
  <div>
    <div>
      Janis
    </div>
    <div>
      Berzins
    </div>
    <div>
      10
    </div>
    <div>
      mail@mail.ll
    </div>
    <div>
      Some company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Peteris
    </div>
    <div>
      Klavins
    </div>
    <div>
      20
    </div>
    <div>
      rail@mail.ll
    </div>
    <div>
      One company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Zane
    </div>
    <div>
      Kocina
    </div>
    <div>
      25
    </div>
    <div>
      nail@mail.ll
    </div>
    <div>
      Second company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`Table > Table *Check callback for update data 1`] = `
<div>
  <div>
    <button>
      Add New Row
    </button>
  </div>
  <div>
    <div>
      name
    </div>
    <div>
      surname
    </div>
    <div>
      age
    </div>
    <div>
      email
    </div>
    <div>
      company
    </div>
    <div>
      actions
    </div>
  </div>
  <div>
    <div>
      New name
    </div>
    <div>
      New surname
    </div>
    <div>
      33
    </div>
    <div>
      new@mail.co
    </div>
    <div>
      New company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Peteris
    </div>
    <div>
      Klavins
    </div>
    <div>
      20
    </div>
    <div>
      rail@mail.ll
    </div>
    <div>
      One company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Zane
    </div>
    <div>
      Kocina
    </div>
    <div>
      25
    </div>
    <div>
      nail@mail.ll
    </div>
    <div>
      Second company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`Table > Table *Check callback for update data, incorrect id 1`] = `
<div>
  <div>
    <button>
      Add New Row
    </button>
  </div>
  <div>
    <div>
      name
    </div>
    <div>
      surname
    </div>
    <div>
      age
    </div>
    <div>
      email
    </div>
    <div>
      company
    </div>
    <div>
      actions
    </div>
  </div>
  <div>
    <div>
      Janis
    </div>
    <div>
      Berzins
    </div>
    <div>
      10
    </div>
    <div>
      mail@mail.ll
    </div>
    <div>
      Some company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Peteris
    </div>
    <div>
      Klavins
    </div>
    <div>
      20
    </div>
    <div>
      rail@mail.ll
    </div>
    <div>
      One company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Zane
    </div>
    <div>
      Kocina
    </div>
    <div>
      25
    </div>
    <div>
      nail@mail.ll
    </div>
    <div>
      Second company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`Table > Table *Snapshot, simple render 1`] = `
<div>
  <div>
    <button>
      Add New Row
    </button>
  </div>
  <div>
    <div>
      name
    </div>
    <div>
      surname
    </div>
    <div>
      age
    </div>
    <div>
      email
    </div>
    <div>
      company
    </div>
    <div>
      actions
    </div>
  </div>
  <div>
    <div>
      Janis
    </div>
    <div>
      Berzins
    </div>
    <div>
      10
    </div>
    <div>
      mail@mail.ll
    </div>
    <div>
      Some company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Peteris
    </div>
    <div>
      Klavins
    </div>
    <div>
      20
    </div>
    <div>
      rail@mail.ll
    </div>
    <div>
      One company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
  <div>
    <div>
      Zane
    </div>
    <div>
      Kocina
    </div>
    <div>
      25
    </div>
    <div>
      nail@mail.ll
    </div>
    <div>
      Second company
    </div>
    <div
      class=""
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`Table > Table *Snapshot, simple render, no initial data 1`] = `
<div>
  <div>
    <button>
      Add New Row
    </button>
  </div>
  <div>
    <div>
      name
    </div>
    <div>
      surname
    </div>
    <div>
      age
    </div>
    <div>
      email
    </div>
    <div>
      company
    </div>
    <div>
      actions
    </div>
  </div>
</div>
`;
